### PR TITLE
Automatically inherit workspace fields when running cargo new/init

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

close https://github.com/rust-lang/cargo/issues/10870

One thing to note: we will try to ensure the order of the inherited fields. You can see more at https://github.com/rust-lang/cargo/pull/12069#discussion_r1201098013

### How should we test and review this PR?
- [x] unit tests

### CODE REVIEW

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [inline src/cargo/ops/cargo_new.rs:884]: Still need to format the document and handle the `edition` field.

**Review by hi-rustin — COMMENTED** [review]
Working in process.

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs:884]: When writing tests, [ui tests](https://doc.crates.io/contrib/tests/writing.html#ui-tests) would be a good candidate so we precisely know how the files will be formatted

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:870]: > Doing a single write would likely be more complicated on the workspace side (slightly, I suspect we can just reuse "find workspace" function that I think exists) but seems like it'd be easier to format it as we wish vs editing it afterwards (even if editing is easier because we can assume what we are editing.

Yes. I agree. I will try `find_workspace_root `.

File: tests/testsuite/main.rs
Comment (hi-rustin) [inline tests/testsuite/main.rs:27]: I will move other tests for the `new` command to this mod after this PR merging.

**Review by hi-rustin — COMMENTED** [review]
Self check

File: src/cargo/ops/cargo_new.rs
Comment (Muscraft) [inline src/cargo/ops/cargo_new.rs:883]: Why can't `toml_edit`/`toml` be used to modify the `Cargo.toml`

File: tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml
Comment (Muscraft) [inline tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml:3]: All tests currently specify `foo` before it was created. While this may happen sometimes, there should be tests that use glob syntax as well as a test for what happens when it is not a member of the workspace.

**Review by Muscraft — COMMENTED** [review]
This appears to be happening inside of [`mv`](https://github.com/rust-lang/cargo/blob/5aa92f12003f53dbc5e771d91529f3c1183fe05f/src/cargo/ops/cargo_new.rs#L737), which is used by `cargo new` and `cargo init`. I am not sure if we want to support this for `cargo init`, currently only `cargo new` is specified in #10870. This should probably be brought up and discussed in #10870 before this is merged.

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs:883]: There shouldn't be; we should be modifying the file with `toml_edit` rather than using string interpolation

Also, editing after-the-fact being more difficult was one of the potential problems I raised with doing the two-write approach

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs:870]: But we also have another issue about making sure it is a part of the workspace member.  Is this approach throwaway if we do that?  If so, I'd rather we just do this right rather  than hack something together and then have to go through this process all over again.

File: tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml
Comment (epage) [reply tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml:3]: We should probably switch the tests to globs.

Without `foo` already existing, I believe this would be a broken `Cargo.toml`.  This makes the test brittle for if we make other changes to `cargo new` (unless we already promise to support broken `Cargo.toml` files)

**Comment by epage** [comment]
> This appears to be happening inside of [`mv`](https://github.com/rust-lang/cargo/blob/5aa92f12003f53dbc5e771d91529f3c1183fe05f/src/cargo/ops/cargo_new.rs#L737), which is used by `cargo new` and `cargo init`. I am not sure if we want to support this for `cargo init`, currently only `cargo new` is specified in #10870. This should probably be brought up and discussed in #10870 before this is merged.

It specified `new` because it is easy to overlook one or the other.  Looking at the documentation for `init`, it doesn't seem like we advertise these as slightly different roles, so I think its reasonable to do for both

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:883]: > Why can't `toml_edit`/`toml` be used to modify the `Cargo.toml`



> There shouldn't be; we should be modifying the file with `toml_edit` rather than using string interpolation

But if we use `toml_edit`/`toml` it will other the keys alphabetically. But do we want to keep the name first?

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:870]: > Is this approach throwaway if we do that? If so, I'd rather we just do this right rather than hack something together and then have to go through this process all over again.

The function is extracted. So it is very easy to call it after we solve that issue.

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs:883]: `toml` is dependent on the data structure used.

`toml_edit` is format preserving and is what `cargo add` and `cargo rm` use.

File: tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml
Comment (hi-rustin) [reply tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml:3]: > Without `foo` already existing, I believe this would be a broken `Cargo.toml`. This makes the test brittle for if we make other changes to `cargo new` (unless we already promise to support broken `Cargo.toml` files)

It is not a broken `Cargo.toml`, it is a normal `Cargo.toml`. But it will cause a warning that reminds the user put the `foo` package into the members or exclude it.

Comment (hi-rustin) [reply tests/testsuite/cargo_new/inherit_workspace_package_table.in/Cargo.toml]: Tests updated and add a new test for the new package is not a workspace member.

**Comment by hi-rustin** [comment]
> It specified `new` because it is easy to overlook one or the other. Looking at the documentation for `init`, it doesn't seem like we advertise these as slightly different roles, so I think its reasonable to do for both

Added a test for it.

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:883]: > `toml` is dependent on the data structure used.
> 
> `toml_edit` is format preserving and is what `cargo add` and `cargo rm` use.

Got it. Thanks for your explanation.

But I think the current approach is more reusable and easier. In the future, if we want to add the package to the members atomically, we only need to call `create_manifest_with_inherited_workspace_package_fields`. If we use `toml_edit` to edit the existing file, it will make the second write couple with the first write. But maybe in the future, we will remove the first write and only keep the second one.

Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Updated the comment.

**Comment by hi-rustin** [comment]
@epage Friendly ping~

Do you have time to take another look? Thanks!

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs]: I consider using `toml_edit` a blocker for merging.

Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: I will give it a try:(

Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Updated!
@epage Do you think it would be beneficial to use 'toml_edit' when creating a new 'Cargo.toml'?

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs:764]: Tests are failing on this commit.

Please make sure all commits pass tests as its important for us to see what the effect is of a change.  In this case, I would consider the new output a regression

e.g.
```
---- expected: tests/testsuite/init/lib_already_exists_nosrc/out/Cargo.toml
++++ actual:   /home/epage/src/personal/cargo/target/tmp/cit/t1374/case/Cargo.toml
        1 + lib = { name = "case", path = "lib.rs" }
        2 +
   1    3 | [package]
   2    4 | name = "case"
   3    5 | version = "0.1.0"
   4    6 | edition = "2021"
   5    7 |
   6    8 | # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
   7    9 |
   8   10 | [dependencies]
   9      -
  10      - [lib]
  11      - name = "case"
  12      - path = "lib.rs"
tests/testsuite/init/lib_already_exists_nosrc/out/lib.rs: is good
```

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs:790]: Why `or_insert`?  you know there is no `bin` key

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs]: This looks like a refactor of the previous commit and unrelated to this commit.  Could you isolate the changes so we can more easily see what is related to the intent of the current commit?

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs:801]: Similarly, this looks unrelated to the current commit

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs:811]: As rust-lang/cargo#12148 is being merged before this, could you update this to also include `workspace.lints`?

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs:805]: Should we reuse this below when we do `paths::write`?

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs:937]: Would it work for us to unconditonally remove the package key so we can just have a single loop?

Also, I assume we remove, rather than overwrite, so inherited keys go at the end?  Thats the style I go with but we should probably explicitly state in the PR description stylistic choices like that

File: tests/testsuite/main.rs
Comment (epage) [inline tests/testsuite/main.rs:143]: nit: we probably shouldn't remove this newline

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs:937]: The downside to not fully looping is we are not fully preserving the order of `[workspace.package]`

File: tests/testsuite/cargo_new/inherit_workspace_package_table_without_version/mod.rs
Comment (epage) [inline tests/testsuite/cargo_new/inherit_workspace_package_table_without_version/mod.rs:14]: How come we use `--edition` in this test?

**Comment by epage** [comment]
This is a lot of great progress; thanks!

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:790]: Because this is a for loop.

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Updated.

File: tests/testsuite/cargo_new/inherit_workspace_package_table_without_version/mod.rs
Comment (hi-rustin) [reply tests/testsuite/cargo_new/inherit_workspace_package_table_without_version/mod.rs]: Removed.

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: We need to consider the case where opts are passed. It has a higher priority than inheritance.

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Fixed.

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:801]: Fixed.

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs:790]: Right, the downside of looking too closely at a change and not the context.

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:764]: Fixed.

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs]: That isn't the case for `version`.

For `edition` and `publish`, we could just short-circuit the loop (`if key == "edition" && opts.edition.is_some() { continue; }`_).  That will also make  the skipping for user-provided values more obvious (somehow I had missed it in my first pass and had to re-read multiple times before I picked it up).

If nothing else, when looking at the tests, it seems weird that for almost all the fields the order of `[workspace.package]` is preserved but not these.

File: src/cargo/ops/cargo_new.rs
Comment (epage) [inline src/cargo/ops/cargo_new.rs]: Couldn't we simplify this to `manifest.to_string()`?

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Do you mind if I add it to the follow-up PR? I need some time to check out this new feature. Also, it is an unstable new feature, do we have to support inheriting it too?

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Updated. Sounds good. 
Thanks for your review!  💚 💙 💜 💛 ❤️

File: src/cargo/ops/cargo_new.rs
Comment (epage) [reply src/cargo/ops/cargo_new.rs]: Yes, we can do it in a follow up.  On merging this, we should note it in #12115, both that it needs to happen and when its resolved

It is as minimum a blocker for stabilization of #12115 but it would be preferable for us to not let unstable features fall behind as we do feature and bug fixes.  In this case, there is enough active interest (from me at least) that we'd keep up but in general it cane widen the gulf enough to make it that much harder to stabilize things in the future.  For example, I tested `lints` against `profile-rustflags` and made some tweaks.

**Review by hi-rustin — COMMENTED** [review]
Self check

File: src/cargo/ops/cargo_new.rs
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Sounds good. I will try it in my next PR.

**Review by epage — APPROVED** [review]
Thanks for your work on this and sticking through the more dramatic changes in the approach!  This both better prepares us for future changes and makes it easier to add other rich features to `cargo add` / `cargo init`.

Let me know if this is ready to merge or if you had other updates you were working on.

**Comment by hi-rustin** [comment]
> Let me know if this is ready to merge or if you had other updates you were working on.

I think it is ready to go.

> Thanks for your work on this and sticking through the more dramatic changes in the approach! This both better prepares us for future changes and makes it easier to add other rich features to `cargo add` / `cargo init`.

Thanks for your patient review!  💚 💙 💜 💛 ❤️ 
This approach definitely better than before! Learned a lot.

File: tests/testsuite/main.rs
Comment (hi-rustin) [reply tests/testsuite/main.rs:27]: @epage Do you believe it's necessary to migrate it, or can we retain it?

Comment (epage) [reply tests/testsuite/main.rs:27]: Moving things around are under discussion, so don't worry about it

**Comment by epage** [comment]
FYI @hi-rustin though this didn't make the release blog post, people seemed to be excited by it
- My [post on reddit](https://www.reddit.com/r/rust/comments/14ylty0/announcing_rust_1710/jrt3u9r/) is the most upvoted
- My [post on mastadon](https://hachyderm.io/@epage/110707292330840528) has gotten some attention as well

So far my favorite quote:

> That is a really cool feature! I didn't knew I needed it, but I actually do!

**Pull request merged**
